packages/engine/lib/provider/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+//! #         provider::request_by_type_tag::<'_, tags::Ref<T>, _>(self)
+//! #     }
+//! # }
+//! fn report_error(e: &(dyn MyErrorTrait + 'static)) {
+//!     // Generic error handling
+//!     // ...
+//!
+//!     // print backtrace
+//!     if let Some(backtrace) = e.request_ref::<Backtrace>() {
+//!         println!("{backtrace:?}")
+//!     }
+//!     # assert!(e.request_ref::<Backtrace>().is_some());
+//!
+//!     // print suggestion text
+//!     if let Some(suggestions) = e.request_ref::<str>() {
+//!         println!("Suggestion: {suggestions}")
+//!     }
+//!     # assert_eq!(e.request_ref::<str>().unwrap(), "Do it correctly next time!");
+//! }
+//!
+//! fn main() {
+//!     let error = MyError {
+//!         backtrace: Backtrace::capture(),
+//!         suggestion: "Do it correctly next time!".to_string(),
+//!     };
+//!
+//!     report_error(&error);
+//! }
+//! ```
+
+// Heavily inspired by https://github.com/rust-lang/project-error-handling/issues/3:
+//   The project-error-handling tries to improves the error trait. In order to move the trait into
+//   `core`, an alternative solution to backtrace provisioning had to be found. This is, where the
+//   provider API comes from.
+//
+//   TODO: replace library with https://github.com/rust-lang/project-error-handling/issues/3.
+
+#![warn(clippy::pedantic, clippy::nursery)]
+
+pub mod tags;
+
+mod internal;
+mod requisition;
+
+use core::any::TypeId;
+
+use self::internal::{TagValue, Tagged};
+use crate::requisition::{ConcreteRequisition, RequisitionImpl};
+
+/// Trait implemented by a type which can dynamically provide tagged values.
+pub trait Provider {
+    /// Object providers should implement this method to provide *all* values they are able to
+    /// provide using `req`.
+    fn provide<'p>(&'p self, req: Requisition<'p, '_>);
+}
+
+/// Request a specific value by a given tag from the `Provider`.
+pub fn request_by_type_tag<'p, I, P: Provider + ?Sized>(provider: &'p P) -> Option<I::Type>
+where
+    I: TypeTag<'p>,
+{
+    let mut req: ConcreteRequisition<'p, I> = RequisitionImpl {
+        tagged: TagValue(None),
+    };
+    provider.provide(Requisition(&mut req));
+    req.tagged.0
+}
+
+/// This trait is implemented by specific `TypeTag` types in order to allow describing a type which
+/// can be requested for a given lifetime `'p`.
+///
+/// A few example implementations for type-driven `TypeTag`s can be found in the [`tags`] module,
+/// although crates may also implement their own tags for more complex types with internal
+/// lifetimes.
+pub trait TypeTag<'p>: Sized + 'static {
+    /// The type of values which may be tagged by this `TypeTag` for the given lifetime.
+    type Type: 'p;
+}
+
+/// A helper object for providing objects by type.
+///
+/// An object provider provides values by calling this type's provide methods. Note, that
+/// `Requisition` is a wrapper around a mutable reference to a [`TypeTag`]ged value.
+#[allow(missing_debug_implementations)]
```

Comment:
Do we need this? A quick google implied to me that the lint is for internal Rust stuff like in std. Assuming it's a remnant of copying part of the Provider impl, but if it's a lint that would appear for us fair enough.